cranelift/codegen/src/isa/x64/inst/emit.rs

```diff
@@ -1906,23 +1913,54 @@ pub(crate) fn emit(
             sink.bind_label(done);
         }
 
-        Inst::XmmRmRImm { op, src, dst, imm } => {
-            let prefix = match op {
-                SseOpcode::Cmpps => LegacyPrefixes::None,
-                SseOpcode::Cmppd => LegacyPrefixes::_66,
-                SseOpcode::Cmpss => LegacyPrefixes::_F3,
-                SseOpcode::Cmpsd => LegacyPrefixes::_F2,
+        Inst::XmmRmRImm {
+            op,
+            src,
+            dst,
+            imm,
+            w,
+        } => {
+            let (prefix, opcode, len) = match op {
+                SseOpcode::Cmpps => (LegacyPrefixes::None, 0x0FC2, 2),
+                SseOpcode::Cmppd => (LegacyPrefixes::_66, 0x0FC2, 2),
+                SseOpcode::Cmpss => (LegacyPrefixes::_F3, 0x0FC2, 2),
+                SseOpcode::Cmpsd => (LegacyPrefixes::_F2, 0x0FC2, 2),
+                SseOpcode::Insertps => (LegacyPrefixes::_66, 0x0F3A21, 3),
+                SseOpcode::Pinsrb => (LegacyPrefixes::_66, 0x0F3A20, 3),
+                SseOpcode::Pinsrw => (LegacyPrefixes::_66, 0x0FC4, 2),
+                SseOpcode::Pinsrd => (LegacyPrefixes::_66, 0x0F3A22, 3),
+                SseOpcode::Pextrb => (LegacyPrefixes::_66, 0x0F3A14, 3),
+                SseOpcode::Pextrw => (LegacyPrefixes::_66, 0x0FC5, 2),
+                SseOpcode::Pextrd => (LegacyPrefixes::_66, 0x0F3A16, 3),
+                SseOpcode::Pshufd => (LegacyPrefixes::_66, 0x0F70, 2),
                 _ => unimplemented!("Opcode {:?} not implemented", op),
             };
-            let opcode = 0x0FC2;
-            let rex = RexFlags::clear_w();
+            let rex = if *w {
+                RexFlags::set_w()
+            } else {
+                RexFlags::clear_w()
+            };
             match src {
                 RegMem::Reg { reg } => {
-                    emit_std_reg_reg(sink, prefix, opcode, 2, dst.to_reg(), *reg, rex);
+                    if *op == SseOpcode::Pextrb || *op == SseOpcode::Pextrd {
+                        // These opcodes (and not the SSE2 version of PEXTRW) flip the operand
+                        // encoding: `dst` in ModRM's r/m, `src` in ModRM's reg field.
+                        emit_std_reg_reg(sink, prefix, opcode, len, *reg, dst.to_reg(), rex);
+                    } else {
+                        // The rest of the opcodes have the customary encoding: `dst` in ModRM's reg,
+                        // `src` in ModRM's r/m field.
+                        emit_std_reg_reg(sink, prefix, opcode, len, dst.to_reg(), *reg, rex);
+                    }
                 }
                 RegMem::Mem { addr } => {
                     let addr = &addr.finalize(state);
-                    emit_std_reg_mem(sink, prefix, opcode, 2, dst.to_reg(), addr, rex);
+                    if *op == SseOpcode::Pextrb || *op == SseOpcode::Pextrd {
```

Comment:
I'd prefer an `assert!()` here to ensure we avoid those opcodes, then avoid the extra nesting level of the if/else.

Reply:
Even better, for static goodness: use a different Vcode instruction for these two opcodes, so that the error becomes a compile time error.

Reply:
I went with the `assert!` here; it's a bit burdensome to create new `Inst` variants and creating new ones just adds to the chaos of deciding which variant to pick during lowering.